source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+      // Optional: set the starting phase (otherwise it would start at 0)
+      set_phase transmit(%1), pi;
+   }
+
+Frames track the set of (frequency, phase, scale) throughout the program and 
+are manipulated using the following `set_` and `shift_` commands. Then when
+a pulse is played the carrier signal of the frame is mixed with the pulse itself
+to create the final signal which will be applied on the channel.
+
+Frequency has no default since it must be set at the beginning. Phase defaults
+to 0 and is a value of type ``angle``. Scale defaults to 1.0 and is a value of
+type ``float``. The exact precision of these three parameters is hardware
+specific.
+
+frameof command
+~~~~~~~~~~~~~~~
+
+The ``frame`` type is a compile-time type represnting a frame. It is useful
+when playing a pulse using the frame of a different channel.
+
+A value of type ``frame`` is retrieved by calling the ``frameof`` command
+and passing in a channel.
+
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase, scale) of a frame can be manipulated throughout program
+execution with the following instructions. After a call to one of these
+instructions, future references to that frame will have the new value.
+
+All of the following instructions have the same form. The first parameter can 
+either be a ``frame`` type or a ``channel`` type (in which case the 
+corresponding frame will be used). The second parameter is the value to shift
+or set. It must be the appropriate type, ``float`` for frequency, ``angle`` for
+phase, and ``float`` for scale. Again, the exact precision of these calculations
+is hardware specific.
+
+The relevant instructions are:
+
+- ``set_frequency`` and ``shift_frequency``
+- ``set_phase`` and ``shift_phase``
+- ``set_scale`` and ``shift_scale``
+
+Here's an example of using ``shift_phase`` to calibrate an ``rz`` gate:
+
+.. code-block:: none
+
+   // Shift phase of qubit 0's frame by pi/4, eg. an rz gate with angle -pi/4
+   shift_phase transmit(%0), pi/4;
+
+   // Equivalent, but more verbose
+   frame fr = frameof(transmit(%0));
+   shift_phase fr, pi/4;
+
+   // Define a calibration for the rz gate on all physical qubits
+   defcal rz(angle[20]:theta) %q {
+     shift_phase transmit(%q), -theta;
+   }
+
+Here's an example qubit spectroscopy experiment.
+
+.. code-block:: none
+
+   qubit q;
+
+   const shots = 1000;
+   const dfreq = 1e6; # 1MHz per point
+   const points = 50; # Sweep over 50MHz
+
+   complex[32] iq, average;
+   complex[32] output[points];
+
+   for p in [0 : points-1] {
+     average = 0;
+     for i in [0 : shots-1] {
+       // Assumes suitable calibrations for reset, x, and measure_iq
+       reset q;
+       x q;
+       measure_iq q -> iq;
+
+       average = (average * i + iq) / (i + 1);
+     }
+     shift q;
+     output[p] = average;
+   }
+
+   defcal prelude {
+      set_frequency transmit(%q) 5e9;
+   }
+
+   defcal shift %q {
+     shift_frequency transmit(%q) dfreq;
+   }
```

Comment:
This doesn't seem right to me -- the point of a `defcal` is to "specify a microcoded implementation of a `gate`, `measure`, or `reset` instruction", not to define a new OpenQASM-level symbol that expands into OpenPulse instructions. That being said, what is the right way to do spectroscopy then? Stuffing the `shift_frequency` call in a `defcal` for `reset` or `measure` feels like the wrong solution, but the only other thing I can think of atm requires breaking your requirement that OpenPulse instructions exist only in `defcal` blocks.

Reply:
+1 this feels wrong, but I don't see a solution around this. Perhaps spectroscopy is an edge-case that we have to hack into the circuit model? A similar case worth exploring is how do we express a Rabi experiment?